File: src/scene/encounter/RandomEncounterTable.java
package scene.encounter;

import java.util.List;
import java.util.Random;

/**
 * This provides a probability table for random encounters during a trail scene
 * as well as handling probability shifts.
 * @author Null && Void
 */
public class RandomEncounterTable {
	
	private List<Encounter> encounters;
	private Random random;
	private int maxValue;
	
	/**
	 * Create the RandomEncounterTable so it may be polled for encounters.
	 * @param encounters An array of generated encounters
	 */
	public RandomEncounterTable(List<Encounter> encounters){
		random = new Random();
		this.encounters = encounters;
	}
		
	/**
	 * Find which Encounter is going to be executed, execute that encounter,
	 * and return a Notification/Scene.
	 * @return The EncounterNotification relating to the just executed Encounter
	 */
	public EncounterNotification getRandomEncounter() {

		maxValue = 0;
		for(Encounter encounter : encounters) {
			maxValue += encounter.getValue();
		}
		
		int roll = random.nextInt(maxValue);
		for (Encounter encounter : encounters) {
			System.out.println(encounter + " Value: " + encounter.getValue());
			if (encounter instanceof NullEncounter) {
				for(Encounter other : encounters) {
					other.increaseValue(1);
				}
				return encounter.doEncounter();
			} else if(encounter.isInRange(roll)) {
				return encounter.doEncounter();
			} else 
				roll -= encounter.getValue();
		}
		return null;
	}
}

File: web/src/scene/encounter/RandomEncounterTable.ts
import { Party } from "../../model/Party";
import { Encounter } from "./Encounter";
import { EncounterNotification } from "./EncounterNotification";
import { EncounterID } from "./EncounterID";
import { NullEncounter } from "./NullEncounter";
import { ThiefEncounter } from "./ThiefEncounter";
import { ItemEncounter } from "./ItemEncounter";
import { PotholeEncounter } from "./PotholeEncounter";
import { MessageEncounter } from "./MessageEncounter";
import { RiverEncounter } from "./RiverEncounter";

export class RandomEncounterTable {
  private encounters: Encounter[];
  private maxValue: number;

  constructor(encounters: Encounter[]) {
    this.encounters = encounters;
    this.maxValue = 0;
  }

  getRandomEncounter(timeIndex: number): EncounterNotification | null {
    this.maxValue = 0;
    for (const encounter of this.encounters) {
      this.maxValue += encounter.getValue(timeIndex);
    }

    let roll = Math.floor(Math.random() * this.maxValue);

    for (const encounter of this.encounters) {
      console.log(`${encounter} Value: ${encounter.getValue(timeIndex)}`);

      if (encounter instanceof NullEncounter) {
        for (const other of this.encounters) {
          other.increaseValue(1);
        }
        return encounter.doEncounter();
      } else if (encounter.isInRange(roll, timeIndex)) {
        return encounter.doEncounter();
      } else {
        roll -= encounter.getValue(timeIndex);
      }
    }

    return null;
  }

  static createEncounters(party: Party): Encounter[] {
    return [
      new ThiefEncounter(party, 1),
      new ItemEncounter(party, 1),
      new PotholeEncounter(party, 1),
      new MessageEncounter(party, 1),
      new RiverEncounter(party, 1),
      new NullEncounter(party, 1),
    ];
  }
}

export function createEncounter(
  party: Party,
  id: EncounterID,
  value: number,
): Encounter {
  switch (id) {
    case EncounterID.THIEF:
      return new ThiefEncounter(party, value);
    case EncounterID.ITEM:
      return new ItemEncounter(party, value);
    case EncounterID.POTHOLE:
      return new PotholeEncounter(party, value);
    case EncounterID.MESSAGE:
      return new MessageEncounter(party, value);
    case EncounterID.RIVER:
      return new RiverEncounter(party, value);
    case EncounterID.NULL:
    default:
      return new NullEncounter(party, value);
  }
}
